In [1]:
from pathlib import Path
import pandas as pd

In [2]:
flights = pd.read_csv("flights.csv")
airlines = pd.read_csv("airlines.csv")
airports = pd.read_csv("airports.csv")
planes = pd.read_csv("planes.csv")
weather = pd.read_csv("weather.csv")

In [31]:
import matplotlib.pyplot as plt

In [3]:
airports[airports['IATA_CODE']=='EWR']['CITY'].unique()

array(['Newark'], dtype=object)

In [4]:
flights.head()

year  month  day  dep_time  dep_delay  arr_time  arr_delay carrier tailnum  \
0  2013      1    1     517.0        2.0     830.0       11.0      UA  N14228   
1  2013      1    1     533.0        4.0     850.0       20.0      UA  N24211   
2  2013      1    1     542.0        2.0     923.0       33.0      AA  N619AA   
3  2013      1    1     544.0       -1.0    1004.0      -18.0      B6  N804JB   
4  2013      1    1     554.0       -6.0     812.0      -25.0      DL  N668DN   

   flight origin dest  air_time  distance  hour  minute  
0    1545    EWR  IAH     227.0      1400   5.0    17.0  
1    1714    LGA  IAH     227.0      1416   5.0    33.0  
2    1141    JFK  MIA     160.0      1089   5.0    42.0  
3     725    JFK  BQN     183.0      1576   5.0    44.0  
4     461    LGA  ATL     116.0       762   5.0    54.0

In [5]:
planes.head()

tailnum    year                     type      manufacturer      model  \
0  N10156  2004.0  Fixed wing multi engine           EMBRAER  EMB-145XR   
1  N102UW  1998.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
2  N103US  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
3  N104UW  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
4  N10575  2002.0  Fixed wing multi engine           EMBRAER  EMB-145LR   

   engines  seats  speed     engine  
0        2     55    NaN  Turbo-fan  
1        2    182    NaN  Turbo-fan  
2        2    182    NaN  Turbo-fan  
3        2    182    NaN  Turbo-fan  
4        2     55    NaN  Turbo-fan

# how many total number of days does the flights table cover?
    

In [6]:

total_days = 0
for month in flights['month'].unique():
    total_days += flights[flights['month']== month]['day'].unique()[-1]
print(total_days)

365


# how many departure cities (not airports) does the flights database cover?

In [7]:
airports.head()

IATA_CODE                              AIRPORT         CITY STATE COUNTRY  \
0       ABE  Lehigh Valley International Airport    Allentown    PA     USA   
1       ABI             Abilene Regional Airport      Abilene    TX     USA   
2       ABQ    Albuquerque International Sunport  Albuquerque    NM     USA   
3       ABR            Aberdeen Regional Airport     Aberdeen    SD     USA   
4       ABY   Southwest Georgia Regional Airport       Albany    GA     USA   

   LATITUDE  LONGITUDE  
0  40.65236  -75.44040  
1  32.41132  -99.68190  
2  35.04022 -106.60919  
3  45.44906  -98.42183  
4  31.53552  -84.19447

In [8]:
lst = []
for origin in flights['origin'].unique():
    if airports[airports['IATA_CODE']== origin]['CITY'].unique() not in lst:
        lst.append(airports[airports['IATA_CODE']== origin]['CITY'].unique())
print(len(lst))

2


# what is the relationship between flights and planes tables?

In [29]:
flights.head()

year  month  day  dep_time  dep_delay  arr_time  arr_delay carrier tailnum  \
0  2013      1    1     517.0        2.0     830.0       11.0      UA  N14228   
1  2013      1    1     533.0        4.0     850.0       20.0      UA  N24211   
2  2013      1    1     542.0        2.0     923.0       33.0      AA  N619AA   
3  2013      1    1     544.0       -1.0    1004.0      -18.0      B6  N804JB   
4  2013      1    1     554.0       -6.0     812.0      -25.0      DL  N668DN   

   flight origin dest  air_time  distance  hour  minute  
0    1545    EWR  IAH     227.0      1400   5.0    17.0  
1    1714    LGA  IAH     227.0      1416   5.0    33.0  
2    1141    JFK  MIA     160.0      1089   5.0    42.0  
3     725    JFK  BQN     183.0      1576   5.0    44.0  
4     461    LGA  ATL     116.0       762   5.0    54.0

In [37]:
planes.head()

tailnum    year                     type      manufacturer      model  \
0  N10156  2004.0  Fixed wing multi engine           EMBRAER  EMB-145XR   
1  N102UW  1998.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
2  N103US  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
3  N104UW  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
4  N10575  2002.0  Fixed wing multi engine           EMBRAER  EMB-145LR   

   engines  seats  speed     engine  
0        2     55    NaN  Turbo-fan  
1        2    182    NaN  Turbo-fan  
2        2    182    NaN  Turbo-fan  
3        2    182    NaN  Turbo-fan  
4        2     55    NaN  Turbo-fan

In [51]:
planes['tailnum'].value_counts()

N180US    1
N463WN    1
N12136    1
N945DL    1
N526AS    1
         ..
N333NW    1
N835AS    1
N14179    1
N829UA    1
N900WN    1
Name: tailnum, Length: 3322, dtype: int64

In [52]:
flights['tailnum'].value_counts()

N725MQ    575
N722MQ    513
N723MQ    507
N711MQ    486
N713MQ    483
         ... 
N978SW      1
N78003      1
N910DN      1
N929DN      1
N815AW      1
Name: tailnum, Length: 4043, dtype: int64

In [9]:
flights.corrwith(planes,axis =0)

year        NaN
air_time    NaN
arr_delay   NaN
arr_time    NaN
day         NaN
dep_delay   NaN
dep_time    NaN
distance    NaN
engines     NaN
flight      NaN
hour        NaN
minute      NaN
month       NaN
seats       NaN
speed       NaN
dtype: float64

# which airplane manufacturer incurred the most delays in the analysis period?

In [10]:
planes.head()

tailnum    year                     type      manufacturer      model  \
0  N10156  2004.0  Fixed wing multi engine           EMBRAER  EMB-145XR   
1  N102UW  1998.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
2  N103US  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
3  N104UW  1999.0  Fixed wing multi engine  AIRBUS INDUSTRIE   A320-214   
4  N10575  2002.0  Fixed wing multi engine           EMBRAER  EMB-145LR   

   engines  seats  speed     engine  
0        2     55    NaN  Turbo-fan  
1        2    182    NaN  Turbo-fan  
2        2    182    NaN  Turbo-fan  
3        2    182    NaN  Turbo-fan  
4        2     55    NaN  Turbo-fan

In [11]:
flights.head()

year  month  day  dep_time  dep_delay  arr_time  arr_delay carrier tailnum  \
0  2013      1    1     517.0        2.0     830.0       11.0      UA  N14228   
1  2013      1    1     533.0        4.0     850.0       20.0      UA  N24211   
2  2013      1    1     542.0        2.0     923.0       33.0      AA  N619AA   
3  2013      1    1     544.0       -1.0    1004.0      -18.0      B6  N804JB   
4  2013      1    1     554.0       -6.0     812.0      -25.0      DL  N668DN   

   flight origin dest  air_time  distance  hour  minute  
0    1545    EWR  IAH     227.0      1400   5.0    17.0  
1    1714    LGA  IAH     227.0      1416   5.0    33.0  
2    1141    JFK  MIA     160.0      1089   5.0    42.0  
3     725    JFK  BQN     183.0      1576   5.0    44.0  
4     461    LGA  ATL     116.0       762   5.0    54.0

In [12]:
airlines.head()

IATA_CODE                 AIRLINE
0        UA   United Air Lines Inc.
1        AA  American Airlines Inc.
2        US         US Airways Inc.
3        F9  Frontier Airlines Inc.
4        B6         JetBlue Airways

In [13]:
flights.tail()

year  month  day  dep_time  dep_delay  arr_time  arr_delay carrier  \
336771  2013      9   30       NaN        NaN       NaN        NaN      9E   
336772  2013      9   30       NaN        NaN       NaN        NaN      9E   
336773  2013      9   30       NaN        NaN       NaN        NaN      MQ   
336774  2013      9   30       NaN        NaN       NaN        NaN      MQ   
336775  2013      9   30       NaN        NaN       NaN        NaN      MQ   

       tailnum  flight origin dest  air_time  distance  hour  minute  
336771     NaN    3393    JFK  DCA       NaN       213   NaN     NaN  
336772     NaN    3525    LGA  SYR       NaN       198   NaN     NaN  
336773  N535MQ    3461    LGA  BNA       NaN       764   NaN     NaN  
336774  N511MQ    3572    LGA  CLE       NaN       419   NaN     NaN  
336775  N839MQ    3531    LGA  RDU       NaN       431   NaN     NaN

In [14]:
flights['dep_time'].unique()

array([517., 533., 542., ..., 236., 208., 206.])

In [15]:
flights[['origin','dest']].value_counts()

origin  dest
JFK     LAX     11262
LGA     ATL     10263
        ORD      8857
JFK     SFO      8204
LGA     CLT      6168
                ...  
JFK     STL         1
EWR     LGA         1
LGA     LEX         1
JFK     MEM         1
        BHM         1
Length: 224, dtype: int64

# which are the two most connected cities?

In [11]:
flights.head()

year  month  day  dep_time  dep_delay  arr_time  arr_delay carrier tailnum  \
0  2013      1    1     517.0        2.0     830.0       11.0      UA  N14228   
1  2013      1    1     533.0        4.0     850.0       20.0      UA  N24211   
2  2013      1    1     542.0        2.0     923.0       33.0      AA  N619AA   
3  2013      1    1     544.0       -1.0    1004.0      -18.0      B6  N804JB   
4  2013      1    1     554.0       -6.0     812.0      -25.0      DL  N668DN   

   flight origin dest  air_time  distance  hour  minute  
0    1545    EWR  IAH     227.0      1400   5.0    17.0  
1    1714    LGA  IAH     227.0      1416   5.0    33.0  
2    1141    JFK  MIA     160.0      1089   5.0    42.0  
3     725    JFK  BQN     183.0      1576   5.0    44.0  
4     461    LGA  ATL     116.0       762   5.0    54.0

In [17]:
flights['origin'].value_counts()

EWR    120835
JFK    111279
LGA    104662
Name: origin, dtype: int64

In [24]:
flights[['origin', 'dest']].value_counts()

origin  dest
JFK     LAX     11262
LGA     ATL     10263
        ORD      8857
JFK     SFO      8204
LGA     CLT      6168
                ...  
JFK     STL         1
EWR     LGA         1
LGA     LEX         1
JFK     MEM         1
        BHM         1
Length: 224, dtype: int64

**JFK  LAX** The max connected city - 11262

**LGA ATL**  The 2nd most connected city - 10263

In [16]:
flights['dest'].unique()

array(['IAH', 'MIA', 'BQN', 'ATL', 'ORD', 'FLL', 'IAD', 'MCO', 'PBI',
       'TPA', 'LAX', 'SFO', 'DFW', 'BOS', 'LAS', 'MSP', 'DTW', 'RSW',
       'SJU', 'PHX', 'BWI', 'CLT', 'BUF', 'DEN', 'SNA', 'MSY', 'SLC',
       'XNA', 'MKE', 'SEA', 'ROC', 'SYR', 'SRQ', 'RDU', 'CMH', 'JAX',
       'CHS', 'MEM', 'PIT', 'SAN', 'DCA', 'CLE', 'STL', 'MYR', 'JAC',
       'MDW', 'HNL', 'BNA', 'AUS', 'BTV', 'PHL', 'STT', 'EGE', 'AVL',
       'PWM', 'IND', 'SAV', 'CAK', 'HOU', 'LGB', 'DAY', 'ALB', 'BDL',
       'MHT', 'MSN', 'GSO', 'CVG', 'BUR', 'RIC', 'GSP', 'GRR', 'MCI',
       'ORF', 'SAT', 'SDF', 'PDX', 'SJC', 'OMA', 'CRW', 'OAK', 'SMF',
       'TUL', 'TYS', 'OKC', 'PVD', 'DSM', 'PSE', 'BHM', 'CAE', 'HDN',
       'BZN', 'MTJ', 'EYW', 'PSP', 'ACK', 'BGR', 'ABQ', 'ILM', 'MVY',
       'SBN', 'LEX', 'CHO', 'TVC', 'ANC', 'LGA'], dtype=object)

In [17]:
flights['dest'].value_counts()

ORD    17283
ATL    17215
LAX    16174
BOS    15508
MCO    14082
       ...  
MTJ       15
SBN       10
ANC        8
LEX        1
LGA        1
Name: dest, Length: 105, dtype: int64

In [18]:
flights[(flights['origin']=='EWR')&(flights['dest']=='ORD')]
#ewr = Newark
#LGA = New York
#JFK = New York
flights.tail()

year  month  day  dep_time  dep_delay  arr_time  arr_delay carrier  \
336771  2013      9   30       NaN        NaN       NaN        NaN      9E   
336772  2013      9   30       NaN        NaN       NaN        NaN      9E   
336773  2013      9   30       NaN        NaN       NaN        NaN      MQ   
336774  2013      9   30       NaN        NaN       NaN        NaN      MQ   
336775  2013      9   30       NaN        NaN       NaN        NaN      MQ   

       tailnum  flight origin dest  air_time  distance  hour  minute  
336771     NaN    3393    JFK  DCA       NaN       213   NaN     NaN  
336772     NaN    3525    LGA  SYR       NaN       198   NaN     NaN  
336773  N535MQ    3461    LGA  BNA       NaN       764   NaN     NaN  
336774  N511MQ    3572    LGA  CLE       NaN       419   NaN     NaN  
336775  N839MQ    3531    LGA  RDU       NaN       431   NaN     NaN